libsolidity/analysis/FunctionCallGraph.cpp

```diff
@@ -0,0 +1,337 @@
+	ContractDefinition const& _contract,
+	FunctionCallGraphBuilder::ContractCallGraph const& _creationGraph
+)
+{
+	solAssert(&_creationGraph.contract == &_contract, "");
+
+	FunctionCallGraphBuilder builder(_contract);
+
+	auto getSecondElement = [](auto const& _tuple){ return get<1>(_tuple); };
+
+	// Create graph for all publicly reachable functions
+	for (FunctionTypePointer functionType: _contract.interfaceFunctionList() | views::transform(getSecondElement))
+	{
+		auto const* function = dynamic_cast<FunctionDefinition const*>(&functionType->declaration());
+		auto const* variable = dynamic_cast<VariableDeclaration const*>(&functionType->declaration());
+
+		if (function)
+			builder.functionReferenced(*function);
+		else
+			// If it's not a function, it must be a getter of a public variable; we ignore those
+			solAssert(variable, "");
+	}
+
+	if (_contract.fallbackFunction())
+		builder.functionReferenced(*_contract.fallbackFunction());
+
+	if (_contract.receiveFunction())
+		builder.functionReferenced(*_contract.receiveFunction());
+
+	// All functions present in internal dispatch at creation time could potentially be pointers
+	// assigned to state variables and as such may be reachable at deployment time as well.
+	builder.m_currentNode = SpecialNode::InternalDispatch;
+	for (Node const& dispatchTarget: valueOrDefault(_creationGraph.edges, SpecialNode::InternalDispatch, {}))
+	{
+		solAssert(!holds_alternative<SpecialNode>(dispatchTarget), "");
+		solAssert(get<CallableDeclaration const*>(dispatchTarget) != nullptr, "");
+		solAssert(dynamic_cast<CallableDeclaration const*>(get<CallableDeclaration const*>(dispatchTarget)) != nullptr, "");
+
+		// Visit the callable to add not only it but also everything it calls too
+		builder.functionReferenced(*get<CallableDeclaration const*>(dispatchTarget), false);
+	}
+
+	builder.m_currentNode = SpecialNode::Entry;
+	builder.processQueue();
+
+	return make_unique<FunctionCallGraphBuilder::ContractCallGraph>(std::move(builder.m_graph));
+}
+
+bool FunctionCallGraphBuilder::visit(FunctionCall const& _functionCall)
+{
+	solAssert(holds_alternative<SpecialNode>(m_currentNode) || get<CallableDeclaration const*>(m_currentNode) != nullptr, "");
+
+	if (*_functionCall.annotation().kind != FunctionCallKind::FunctionCall)
+		return true;
+
+	auto const* functionType = dynamic_cast<FunctionType const*>(_functionCall.expression().annotation().type);
+	solAssert(functionType, "");
+
+	// TODO: Is this actually guaranteed?
+	solAssert(functionType->kind() != FunctionType::Kind::Declaration, "");
+
+	if (functionType->kind() == FunctionType::Kind::Internal && !_functionCall.expression().annotation().calledDirectly)
+		// If it's not a direct call, we don't really know which function will be called (it may even
+		// change at runtime). All we can do is to add an edge to the dispatch which in turn has
+		// edges to all functions could possibly be called.
+		add(m_currentNode, SpecialNode::InternalDispatch);
```

Comment:
We talked about it with @chriseth and it's like that because it does not really matter for events so no one cares about what `requiredLookup` says.

I think I'll submit a small PR later to change this but this is not too important.